## **Introduction**
Here we build our ranking system. We only consider `income_total` which is `rate` * `dollar_value`  as it is a strong indicator for Merchant Performance. 

**Assumption**: `dollar_value` is *Normally Distributed*

Let's load the data after removing fraud transactions.

In [3]:
from pyspark.sql import SparkSession, functions as F
import pandas as pd

# Create a spark session
spark = (
    SparkSession.builder.appName("BNPL Project")
    .config("spark.sql.repl.eagerEval.enabled", True) 
    .config("spark.sql.parquet.cacheMetadata", "true")
    .config("spark.driver.memory", "4g")
    .config("spark.sql.session.timeZone", "Etc/UTC")
    .getOrCreate()
)

sdf = spark.read.parquet("../data/curated/process_data.parquet/")

cols = ['merchant_name', 'user_id', 'dollar_value', 'order_datetime', 'rate', 'category', 'subcategory', 'estimated_region_population_2021_sum',	'persons_earners_2018-19_sum', 'mean_earnings_2018-19_avg', 'sum_earnings_2018-19_sum',	'median_earnings_2018-19_avg', 'med_age_earners_2018-19_avg']
sdf = sdf.select(cols)

sdf.limit(10)

22/10/09 00:34:59 WARN Utils: Your hostname, Shromanns-MacBook-Pro.local resolves to a loopback address: 127.0.0.1; using 192.168.1.110 instead (on interface en0)
22/10/09 00:34:59 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


22/10/09 00:35:00 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


+--------------------+-------+------------+--------------+------+--------------------+--------------------+------------------------------------+---------------------------+-------------------------+------------------------+---------------------------+---------------------------+
|       merchant_name|user_id|dollar_value|order_datetime|  rate|            category|         subcategory|estimated_region_population_2021_sum|persons_earners_2018-19_sum|mean_earnings_2018-19_avg|sum_earnings_2018-19_sum|median_earnings_2018-19_avg|med_age_earners_2018-19_avg|
+--------------------+-------+------------+--------------+------+--------------------+--------------------+------------------------------------+---------------------------+-------------------------+------------------------+---------------------------+---------------------------+
|Dolor Dapibus Gra...|   8913|       51.28|    2021-07-24|0.0312|retail_and_wholes...|household_goods_r...|                               31499|                    28552.0|                  70738.0|            2.01971729E9|                    28339.0|                       31.0|
|    Lorem Foundation|  15797|        5.84|    2021-11-26|0.0655|retail_and_wholes...|household_goods_r...|                                3266|                     2023.0|                  60415.0|            1.22219664E8|                    49925.0|                       48.0|
|Euismod Urna Inst...|   4694|        0.36|    2022-05-07|0.0505|retail_and_wholes...|   department_stores|                               22545|                    14289.0|                  84223.0|           1.203458028E9|                    54884.0|                       47.0|
|         Mauris Inc.|   3260|        5.72|    2022-04-18|0.0294|info_media_and_te...|                null|                               43353|                    25261.0|                  82029.0|           1.628989984E9|                    54081.0|         44.333333333333336|
|Lobortis Ultrices...|  15618|       69.08|    2022-08-13|0.0631|retail_and_wholes...|   department_stores|                               37268|                    22626.0|       57968.333333333336|           1.310262977E9|         53277.333333333336|                       44.0|
|   Imperdiet Non LLC|  22134|      129.51|    2021-10-11|0.0684|info_media_and_te...|                null|                                7482|                     4086.0|                  43661.0|            1.77649122E8|                    39343.5|                       46.0|
|Eget Venenatis A ...|   8914|      313.78|    2021-04-25|0.0687|retail_and_wholes...|   department_stores|                               97517|                    50162.0|                  63845.8|             3.1775621E9|                    55556.4|                       39.0|
|Non Vestibulum In...|   1604|       21.55|    2021-05-13| 0.058|retail_and_wholes...|   department_stores|                                4296|                     2344.0|                  51198.0|            1.20007795E8|                    43117.0|                       48.0|
|Lobortis Ultrices...|   2786|       60.47|    2022-08-14|0.0631|retail_and_wholes...|   department_stores|                                8448|                     4341.0|                  54531.0|            2.36717978E8|                    46601.0|                       45.0|
|Lobortis Ultrices...|  21018|       56.28|    2021-12-27|0.0631|retail_and_wholes...|   department_stores|                               14916|                     7568.0|                  54270.0|            4.23766641E8|         50905.666666666664|         40.666666666666664|
+--------------------+-------+------------+--------------+------+--------------------+--------------------+------------------------------------+---------------------------+-------------------------+------------------------+---------------------------+---------------------------+

### 1. **Identify** Merchants with the greatest gross and lowest volatility in income and the greatest evidence of performance
We will order the dataset in decreasing value in sum of `total_income` and number of instances and increasing value in `number of instances` per merchant.
This is because we want to identify the the wealthiest merchant to partner with, hence the sum of `total_income`; yet also making sure they earn income on a consistent basis, hence the standard deviation of `total_income` and finally have enough evidence to support their claims, hence the `number of instances`.

In [4]:
fullSdf = sdf.groupBy('merchant_name').agg(
        F.sum('dollar_value').alias('income_total'),
        F.stddev('dollar_value').alias('income_deviation'),
        F.mean('rate').alias('rate'),
        F.count('merchant_name').alias('count_merchant'))\
    .withColumn('income_total', F.col('income_total') * F.col('rate'))\
    .withColumn('income_deviation', F.col('income_deviation') * F.col('rate'))\
    .orderBy(F.col('count_merchant').desc(), F.col('income_total').desc(), F.col('income_deviation').asc())

fullSdf

+--------------------+------------------+------------------+--------------------+--------------+
|       merchant_name|      income_total|  income_deviation|                rate|count_merchant|
+--------------------+------------------+------------------+--------------------+--------------+
|      Erat Vitae LLP| 255744.1228860645|0.6226536011043045|0.029400000000007417|        289513|
|   Leo In Consulting|  613410.486442027|1.5931132977937763| 0.06430000000000283|        272674|
|  Pede Nonummy Corp.| 215914.1209939455|0.4297217319521723|0.028599999999992777|        260780|
|Non Vestibulum In...| 405304.2824599577|0.8686670077990107| 0.05799999999999398|        243520|
|Suspendisse Dui C...| 206349.6233179613|0.8804237642008537|0.029299999999994514|        234397|
|    Lacus Consulting| 585975.3666878997| 2.563138949123425| 0.06979999999998802|        228219|
| Est Nunc Consulting|  535420.132484091|1.4298403330446474|0.060100000000010194|        215963|
|Lorem Ipsum Sodal...| 404176.7518500285| 1.159906434919491| 0.04470000000000315|        200913|
|Ipsum Dolor Sit C...|106133.55576602311|0.5410840739271985|0.014700000000003183|        196816|
|Vehicula Pellente...|285260.53175996774|1.5744092874763471| 0.06479999999999268|        181841|
|Placerat Eget Ven...|212449.64405003347|0.9557664631557532|0.030500000000004815|        180824|
|Mauris Non Institute| 575026.0092899798| 3.355073032852449| 0.06099999999999779|        120624|
|Lobortis Ultrices...|   559913.80276603|2.6244704684791995| 0.06310000000000343|        113541|
|   Nullam Consulting|  451904.946444086|3.1426202633077285| 0.06330000000001205|        101791|
|Phasellus At Company|  451756.970280048|2.8687394923898606| 0.04950000000000524|         99226|
|Dictum Phasellus ...| 515033.4164851049| 3.293324445514841|0.056500000000011485|         99176|
|Ultricies Digniss...|376464.87236396025|1.5864039550422104| 0.04689999999999509|         90705|
|Orci In Consequat...| 623617.3677570462| 4.203807054726122| 0.06610000000000485|         85780|
|Euismod Urna Inst...| 54714.83402999952|  0.70210821882138|0.050499999999999566|         78335|
|Suspendisse Ac As...| 337488.9689199826| 2.856199098591569|0.046999999999997634|         68240|
+--------------------+------------------+------------------+--------------------+--------------+
only showing top 20 rows

In [5]:
days = sdf.select('order_datetime').distinct().count()

### 2. **Ranking** Merchants by Categories

In [16]:
def topNmerchants(sdf, categories, N):
    merchants = {}
    income_total = {}
    income_risk = {}
    counts = {}

    for category in categories:
        tdf = sdf.where(F.col('category') == category)
        tdf = tdf.groupBy('merchant_name').agg(
            F.sum('dollar_value').alias('income_total'),
            F.stddev('dollar_value').alias('income_deviation'),
            F.mean('rate').alias('rate'),
            F.count('merchant_name').alias('count_merchant'))\
        .withColumn('income_total', F.col('income_total') * F.col('rate'))\
        .withColumn('income_deviation', F.col('income_deviation') * F.col('rate'))\
        .orderBy(F.col('count_merchant').desc(), F.col('income_total').desc(), F.col('income_deviation').asc())\
        
        tdf = tdf.limit(N)

        merchants[category] = tdf.select('merchant_name').toPandas()['merchant_name'].to_list()
        
        # assuming that all income is Normally distributed
        income_total[category] = tdf.agg({'income_total': 'sum'}).first()['sum(income_total)']
        income_risk[category] = tdf.agg({'income_deviation': 'sum'}).first()['sum(income_deviation)']
        counts[category] = tdf.agg({'count_merchant': 'sum'}).first()['sum(count_merchant)']

    return pd.DataFrame(merchants), income_total, income_risk, counts

In [17]:
categories = ['retail_and_wholesale_trade', 'rental_hiring_and_real_estate', 'arts_and_recreation', 'info_media_and_telecommunications', 'others']

We see that the top 10 merchants per category is as shown below

In [18]:
merchants, total, risk, counts = topNmerchants(sdf, categories, N=10)
merchants

retail_and_wholesale_trade  rental_hiring_and_real_estate  \
0                     Erat Vitae LLP  Quis Massa Mauris Corporation   
1                  Leo In Consulting      Quam A Felis Incorporated   
2                 Pede Nonummy Corp.             Vel Lectus Cum LLC   
3          Non Vestibulum Industries            Mi Eleifend Company   
4        Suspendisse Dui Corporation             Iaculis Enim Corp.   
5                   Lacus Consulting              Leo Morbi Limited   
6                Est Nunc Consulting                   Laoreet Inc.   
7     Lorem Ipsum Sodales Industries                  Massa Limited   
8        Ipsum Dolor Sit Corporation            Fermentum Institute   
9  Vehicula Pellentesque Corporation                Morbi Non Corp.   

                    arts_and_recreation  info_media_and_telecommunications  \
0                    Ac Urna Consulting               Mauris Non Institute   
1                  Magna Sed Industries                   Euismod In Corp.   
2                         Mi Consulting          Feugiat Sed Nec Institute   
3                             Lorem LLP                    Amet Consulting   
4  Volutpat Ornare Facilisis Associates           Arcu Sed Eu Incorporated   
5            Ligula Tortor Incorporated                        Eleifend PC   
6                Nullam Scelerisque Ltd  Posuere Cubilia Curae Corporation   
7                             Massa LLP                       At Sem Corp.   
8                        Non Cursus LLP                 Et Nunc Consulting   
9            Mollis Duis Sit Foundation           Suspendisse Incorporated   

                         others  
0                Nec Tellus Ltd  
1      Tempus Eu Ligula Limited  
2                  Sed Nec Inc.  
3            Natoque Consulting  
4        Gravida Praesent Corp.  
5         Mauris Sagittis Corp.  
6            Nisl Elementum Ltd  
7     Molestie Arcu Corporation  
8    Feugiat Lorem Incorporated  
9  Risus Odio Auctor Foundation

### a. **Total** revenue made by each category

In [20]:
total

{'retail_and_wholesale_trade': 3613688.9746479657,
 'rental_hiring_and_real_estate': 351859.06383600016,
 'arts_and_recreation': 878824.603607,
 'info_media_and_telecommunications': 2282460.32927698,
 'others': 705908.5849809962}

### b. **Risk** in revenue per category (standard deviation)

In [21]:
risk

{'retail_and_wholesale_trade': 11.66295848134123,
 'rental_hiring_and_real_estate': 322.02846931549146,
 'arts_and_recreation': 650.0662699383139,
 'info_media_and_telecommunications': 38.42206648478936,
 'others': 58.94039668372187}

### c. **Number** of Instances per category

In [22]:
counts

{'retail_and_wholesale_trade': 2324636,
 'rental_hiring_and_real_estate': 18007,
 'arts_and_recreation': 12123,
 'info_media_and_telecommunications': 449654,
 'others': 92071}

## **Review**
From our selection of merchants from all categories, we can calculate our long term daily average merchant income to be as follows with 3 standard deviations.

Since `dollar_value` is assumed to be *Normally Distributed*, the distribution of the `total_income` per category is also *Normally Distributed* with its paramters being the sum of the selected merchant's `dollar_value`.

$$
m_1 \sim N(\mu_{1}, \sigma^2_{1}), \dots, m_{10} \sim N(\mu_{10}, \sigma^2_{10}) \\
\implies \sum_{i=1}^{10}m_i \sim N\left(\sum_{i=1}^{10} \mu_i, \sum_{i=1}^{10} \sigma^2_i\right)
$$

We use this to find the daily average income and our 3 standard deviations as found below,

In [23]:
mu = sum(total.values()) / days
sigma = sum(risk.values()) / days

print('Average Daily Income:')
print(f'({mu - (3 * sigma)} <- 3 std - | {mu} | - 3 std -> {mu + (3 * sigma)})')

Average Daily Income:
(12919.964019581239 <- 3 std - | 12925.316099585712 | - 3 std -> 12930.668179590186)


As seen above, the merchants make about $12925 a day; with $12919 as lower and $12930 upper bound.